{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri, FloatVector, ListVector\n",
    "import pandas as pd\n",
    "from subprocess import call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/ASE/nano_merge_signal_eqtl.csv\"\n",
    "df1_trait_col = \"\"\n",
    "df1_trait = \"ASE\"\n",
    "\n",
    "df2_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/m6A/nano_merge_signal_m6Aqtl.csv\"\n",
    "df2_trait_col = \"m6A_pos_1base\"\n",
    "df2_trait = \"m6A\"\n",
    "\n",
    "gene_bed_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/stability/gene_merge.bed\"\n",
    "outdir = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/colco\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1、传入两个QTL的结果 筛选出所有fdr<0.1的点\n",
    "merge_columns = ['chrom', 'strand', 'snp_pos_1base', 'rsID', 'A1', 'A2', 'MAF']\n",
    "df1_column_names = merge_columns + ([df1_trait_col] if df1_trait_col else []) + ['p_value', 'fdr']\n",
    "df2_column_names = merge_columns + ([df2_trait_col] if df2_trait_col else []) + ['p_value', 'fdr']\n",
    "\n",
    "df1 = pd.read_csv(df1_file, usecols = df1_column_names)\n",
    "df1 = df1[df1['fdr'] < 0.1]\n",
    "df2 = pd.read_csv(df2_file, usecols = df2_column_names)\n",
    "df2 = df2[df2['fdr'] < 0.1]\n",
    "\n",
    "del df1['fdr'],df2['fdr']\n",
    "\n",
    "df1 = df1.rename(columns={'p_value': f'p_value_{df1_trait}'})\n",
    "df2 = df2.rename(columns={'p_value': f'p_value_{df2_trait}'})\n",
    "\n",
    "## 2、筛选出所有重叠的QTL位点，然后对这些SNP位点进行gene注释\n",
    "merged_df = pd.merge(df1, df2, on=merge_columns)\n",
    "merged_df_anno = merged_df.drop_duplicates(subset=merge_columns)\n",
    "merged_df_anno['snp_pos_0base'] = merged_df_anno['snp_pos_1base']-1\n",
    "merged_df_anno['score'] = 0\n",
    "merged_df_anno = merged_df_anno[['chrom','snp_pos_0base','snp_pos_1base','score','rsID','strand']]\n",
    "merged_df_anno.to_csv(f\"{outdir}/merged_df_anno.bed\", sep=\"\\t\", index=False, header=False)\n",
    "call(f'bedtools intersect -a {outdir}/merged_df_anno.bed -b {gene_bed_file} -s -wa -wb > {outdir}/merged_df_gene_anno.bed',shell=True)\n",
    "merged_df_anno = pd.read_csv(f'{outdir}/merged_df_gene_anno.bed', sep=\"\\t\", header=None,usecols = [0,2,4,5,9])\n",
    "merged_df_anno.columns = ['chrom','snp_pos_1base','rsID','strand','geneID']\n",
    "merged_df = pd.merge(merged_df, merged_df_anno[['geneID'] + ['chrom', 'snp_pos_1base', 'rsID', 'strand']], on=['chrom', 'snp_pos_1base', 'rsID', 'strand'], how='left')\n",
    "call(f'rm {outdir}/merged_df_anno.bed {outdir}/merged_df_gene_anno.bed',shell=True)\n",
    "\n",
    "## 3、将这些SNP按照gene分类，如果gene中这种QTL含量较少 就合并分析区间，得到每个区间的QTL信息\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "假设我现在想分析mQTL和m6AQTL之间的共定位关系 我是应该以甲基化位点+-100kb为分析区间还是以m6A位点+-100kb为分析区间呢？不同区间的选择是否会对最后的结果有影响"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "\n",
    "def calculate_se_manual(X, y):\n",
    "    # 添加常数项\n",
    "    X = sm.add_constant(X)\n",
    "    \n",
    "    # 拟合模型\n",
    "    model = sm.OLS(y, X).fit()\n",
    "    \n",
    "    # 预测值\n",
    "    y_pred = model.predict(X)\n",
    "    \n",
    "    # 计算残差平方和\n",
    "    rss = np.sum((y - y_pred)**2)\n",
    "    \n",
    "    # 计算自变量的离差平方和\n",
    "    x_demeaned = X[:, 1] - np.mean(X[:, 1])\n",
    "    ssx = np.sum(x_demeaned**2)\n",
    "    \n",
    "    # 计算误差方差的估计\n",
    "    n = len(y)\n",
    "    s_squared = rss / (n - 2)\n",
    "    \n",
    "    # 计算SE\n",
    "    se = np.sqrt(s_squared / ssx)\n",
    "    \n",
    "    return se\n",
    "\n",
    "# 使用statsmodels直接获取SE\n",
    "def calculate_se_statsmodels(X, y):\n",
    "    X = sm.add_constant(X)\n",
    "    model = sm.OLS(y, X).fit()\n",
    "    return model.bse[1]  # 返回斜率的SE\n",
    "\n",
    "# 示例使用\n",
    "X = np.random.rand(100, 1)\n",
    "y = 2 * X + 1 + np.random.randn(100, 1) * 0.1\n",
    "\n",
    "se_manual = calculate_se_manual(X, y)\n",
    "se_statsmodels = calculate_se_statsmodels(X, y)\n",
    "\n",
    "print(f\"Manually calculated SE: {se_manual[0]}\")\n",
    "print(f\"Statsmodels SE: {se_statsmodels}\")\n",
    "\n",
    "\n",
    "df[['p_value','beta', 'SE']] = results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 4、对所有的区间进行共定位分析，这里注意会存在有的SNP同时影响多种表型"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 5、整理出所有区间中PH4>0.75的SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input <- merge(eqtl, gwas, by=\"rs_id\", all=FALSE, suffixes=c(\"_eqtl\",\"_gwas\"))\n",
    "result <- coloc.abf(dataset1=list(pvalues=input$pval_nominal_gwas, type=\"cc\", s=0.33, N=50000), dataset2=list(pvalues=input$pval_nominal_eqtl, type=\"quant\", N=10000), MAF=input$maf)\n",
    "\n",
    "need_result=result$results %>% filter(SNP.PP.H4 > 0.95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 启用pandas和R的数据框之间的转换\n",
    "pandas2ri.activate()\n",
    "\n",
    "# 加载R包\n",
    "coloc = importr('coloc')\n",
    "\n",
    "# 创建pandas数据框\n",
    "data1 = pd.DataFrame({\n",
    "    'snp': ['rs1', 'rs2', 'rs3'],\n",
    "    'beta': [0.2, 0.4, 0.6],\n",
    "    'varbeta': [0.02, 0.04, 0.06],\n",
    "    'MAF': [0.1, 0.2, 0.3],\n",
    "    'N': [1000, 1000, 1000],\n",
    "    'type': ['quant', 'quant', 'quant']\n",
    "})\n",
    "\n",
    "data2 = pd.DataFrame({\n",
    "    'snp': ['rs1', 'rs2', 'rs3'],\n",
    "    'beta': [0.3, 0.5, 0.7],\n",
    "    'varbeta': [0.03, 0.05, 0.07],\n",
    "    'MAF': [0.1, 0.2, 0.3],\n",
    "    'N': [1000, 1000, 1000],\n",
    "    'type': ['quant', 'quant', 'quant']\n",
    "})\n",
    "\n",
    "# 将pandas数据框转换为R数据框\n",
    "rdf1 = pandas2ri.py2rpy(data1)\n",
    "rdf2 = pandas2ri.py2rpy(data2)\n",
    "\n",
    "# 运行coloc.abf函数进行共定位分析\n",
    "result = coloc.coloc_abf(dataset1=rdf1, dataset2=rdf2)\n",
    "\n",
    "# 将结果转换回pandas数据框\n",
    "result_df = pandas2ri.rpy2py(result)\n",
    "print(result_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 假设你有一个输入字典input包含pval_nominal_gwas, pval_nominal_eqtl, 和maf\n",
    "input_data = {\n",
    "    'pval_nominal_gwas': [0.01, 0.02, 0.05],  # 示例数据\n",
    "    'pval_nominal_eqtl': [0.03, 0.04, 0.01],  # 示例数据\n",
    "    'maf': [0.1, 0.2, 0.3]                    # 示例数据\n",
    "}\n",
    "\n",
    "# 将输入数据转换为R的向量\n",
    "pval_nominal_gwas = FloatVector(input_data['pval_nominal_gwas'])\n",
    "pval_nominal_eqtl = FloatVector(input_data['pval_nominal_eqtl'])\n",
    "maf = FloatVector(input_data['maf'])\n",
    "\n",
    "# 创建dataset1和dataset2\n",
    "dataset1 = ListVector({\n",
    "    'pvalues': pval_nominal_gwas,\n",
    "    'type': 'cc',\n",
    "    's': 0.33,\n",
    "    'N': 50000\n",
    "})\n",
    "\n",
    "dataset2 = ListVector({\n",
    "    'pvalues': pval_nominal_eqtl,\n",
    "    'type': 'quant',\n",
    "    'N': 10000\n",
    "})\n",
    "\n",
    "# 运行coloc.abf函数\n",
    "result = coloc.coloc_abf(dataset1=dataset1, dataset2=dataset2, MAF=maf)\n",
    "\n",
    "# 将结果转换为Python数据结构（如需要）\n",
    "result_df = pandas2ri.rpy2py(result)\n",
    "print(result_df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "methy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
