{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from subprocess import call\n",
    "import itertools\n",
    "import numpy as np\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import pandas2ri\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ASE_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/ASE/nano_merge_ASE_SNP.csv\"\n",
    "m6A_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/m6A/nano_merge_m6A_SNP.csv\"\n",
    "pseU_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/pseU/nano_merge_pseU_SNP.csv\"\n",
    "stability_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/stability/nano_merge_stability_SNP.csv\"\n",
    "isoform_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/isoform/nano_merge_isoform_SNP.csv\"\n",
    "polyA_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/apa/nano_merge_polyA_SNP.csv\"\n",
    "promoter_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/promoter/nano_merge_promoter_SNP.csv\"\n",
    "tss_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/tss/nano_merge_tss_SNP.csv\"\n",
    "\n",
    "gene_bed_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/stability/gene_merge.bed\"\n",
    "outdir = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.7.2/colco\"\n",
    "\n",
    "chrom = \"chr1\"\n",
    "strand = \"+\"\n",
    "min_qtl_count=25\n",
    "coloc_threshold = 0.75\n",
    "p_threshold=0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_file(chrom, strand, full_summury, trait, trait_col=''):\n",
    "    df = pd.read_csv(full_summury)\n",
    "    df = df[(df['chrom'] == chrom)&(df['strand'] == strand)]\n",
    "    df = df[df['p_value']<1]\n",
    "    df = df.rename(columns={'p_value': f'p_value_{trait}'})\n",
    "    df = df['snp_pos_1base'].astype(str)\n",
    "    common_col = ['rsID', 'chrom', 'snp_pos_1base', 'strand', 'EAF', f'p_value_{trait}']\n",
    "    if trait_col:\n",
    "        common_col.append(trait_col)\n",
    "    df = df[common_col]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31058/2141567306.py:2: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(full_summury)\n",
      "/tmp/ipykernel_31058/2141567306.py:2: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(full_summury)\n"
     ]
    }
   ],
   "source": [
    "## 读所有的文件\n",
    "ASE_df = read_file(chrom, strand, ASE_file, \"ASE\")\n",
    "m6A_df = read_file(chrom, strand, m6A_file, \"m6A\", \"m6A_pos_1base\")\n",
    "pseU_df = read_file(chrom, strand, pseU_file, \"pseU\", \"pseU_pos_1base\")\n",
    "stability_df = read_file(chrom, strand, stability_file, \"stability\")\n",
    "isoform_df = read_file(chrom, strand, isoform_file, \"isoform\")\n",
    "polyA_df = read_file(chrom, strand, polyA_file, \"polyA\")\n",
    "promoter_df = read_file(chrom, strand, promoter_file, \"promoter\")\n",
    "tss_df = read_file(chrom, strand, tss_file, \"tss\")\n",
    "\n",
    "## 合并所有的 DataFrame\n",
    "dfs = [ASE_df, stability_df, isoform_df, polyA_df, promoter_df, tss_df, m6A_df, pseU_df]\n",
    "merged_df = dfs[0]\n",
    "for df in dfs[1:]:\n",
    "    merged_df = pd.merge(merged_df, df, on=['rsID', 'chrom', 'snp_pos_1base', 'strand', 'EAF'], how='outer')\n",
    "merged_df['non_null_count'] = merged_df.notnull().sum(axis=1) # 统计每行的非空值数量\n",
    "merged_df = merged_df[merged_df['non_null_count'] >= 10]  # 4列键 + 至少2种trait(6列)，保证这个SNP至少是两种QTL\n",
    "merged_df = merged_df.drop(columns=['non_null_count'])\n",
    "\n",
    "## gene注释\n",
    "gene_df = merged_df[['chrom', 'snp_pos_1base', 'strand', 'rsID']]\n",
    "gene_df = gene_df.drop_duplicates()\n",
    "gene_df['snp_pos_0base'] = gene_df['snp_pos_1base'] - 1\n",
    "gene_df['score'] = 0\n",
    "gene_df = gene_df[['chrom', 'snp_pos_0base', 'snp_pos_1base', 'rsID', 'score', 'strand']]\n",
    "gene_df.to_csv(f'{outdir}/rs_{chrom}_{strand}.bed', sep='\\t', index=False, header=False)\n",
    "call(f'bedtools intersect -wa -wb -s -a {outdir}/rs_{chrom}_{strand}.bed -b {gene_bed_file} > {outdir}/gene_rs_{chrom}_{strand}.bed',shell=True)\n",
    "gene_df = pd.read_csv(f'{outdir}/gene_rs_{chrom}_{strand}.bed',header=None,sep='\\t',usecols = [0,2,3,5,9],names=['chrom','snp_pos_1base','rsID','strand','geneID'])\n",
    "merged_df = pd.merge(merged_df, gene_df[['chrom', 'snp_pos_1base', 'rsID', 'strand', 'geneID']], on=['chrom', 'snp_pos_1base', 'rsID', 'strand'], how='left')\n",
    "call(f'rm {outdir}/rs_{chrom}_{strand}.bed {outdir}/gene_rs_{chrom}_{strand}.bed',shell=True)\n",
    "\n",
    "## 分区进行colco\n",
    "gene_groups = merged_df.groupby('geneID')\n",
    "for gene_group in gene_groups:\n",
    "    geneID = gene_group[0]\n",
    "    gene_df = gene_group[1]\n",
    "    genes_coloc_res = pd.DataFrame(columns=[\n",
    "        \"chrom\",\"snp_pos_1base\",\"rsID\",\"strand\",\"EAF\",\n",
    "        \"p_value_1\",\"p_value_2\"\n",
    "        \"PPH4\",\"trait1\",\"trait2\"\n",
    "        \"geneID\"\n",
    "        ])\n",
    "    # 筛选1:首先对gene整体筛选 保证gene内至少有min_qtl_count个重叠QTL\n",
    "    if len(set(gene_df['snp_pos_1base'])) >= min_qtl_count:\n",
    "        gene_coloc_res = gene_2trait_combiner(geneID, gene_df, min_qtl_count, p_threshold, coloc_threshold)\n",
    "        if gene_coloc_res:\n",
    "            genes_coloc_res = pd.concat([genes_coloc_res, gene_coloc_res], ignore_index=True)\n",
    "        else:\n",
    "            continue\n",
    "    else:\n",
    "        print(f\"{geneID} has less than {min_qtl_count} shared QTLs, skip\")\n",
    "        continue\n",
    "\n",
    "if len(genes_coloc_res) != 0 :\n",
    "    m6A1 = genes_coloc_res[genes_coloc_res['trait1'] == \"m6A\"]\n",
    "    m6A2 = genes_coloc_res[genes_coloc_res['trait2'] == \"m6A\"]\n",
    "    pseU1 = genes_coloc_res[genes_coloc_res['trait1'] == \"pseU\"]\n",
    "    pseU2 = genes_coloc_res[genes_coloc_res['trait2'] == \"pseU\"]\n",
    "    others = genes_coloc_res[\n",
    "        (genes_coloc_res['trait1'] != \"m6A\") &\n",
    "        (genes_coloc_res['trait1'] != \"pseU\") &\n",
    "        (genes_coloc_res['trait2'] != \"m6A\") &\n",
    "        (genes_coloc_res['trait2'] != \"pseU\")\n",
    "    ]\n",
    "    others = others.assign(m6A_pos_1base='', pseU_pos_1base='')\n",
    "\n",
    "    m6A_df = m6A_df.rename(columns={'p_value_m6A': 'p_value_1'})\n",
    "    m6A1 = pd.merge(\n",
    "        m6A1,\n",
    "        m6A_df[['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'm6A_pos_1base', 'p_value_1']],\n",
    "        on=['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'p_value_1'],\n",
    "        how='left'\n",
    "    )\n",
    "    m6A1['pseU_pos_1base'] = ''\n",
    "\n",
    "    m6A_df = m6A_df.rename(columns={'p_value_1': 'p_value_2'})\n",
    "    m6A2 = pd.merge(\n",
    "        m6A2,\n",
    "        m6A_df[['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'm6A_pos_1base', 'p_value_2']],\n",
    "        on=['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'p_value_2'],\n",
    "        how='left'\n",
    "    )\n",
    "    m6A2['pseU_pos_1base'] = ''\n",
    "\n",
    "\n",
    "    pseU_df = m6A_df.rename(columns={'p_value_pseU': 'p_value_1'})\n",
    "    pseU1 = pd.merge(\n",
    "        pseU1,\n",
    "        pseU_df[['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'pseU_pos_1base', 'p_value_1']],\n",
    "        on=['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'p_value_1'],\n",
    "        how='left'\n",
    "    )\n",
    "    pseU1['m6A_pos_1base'] = ''\n",
    "\n",
    "    pseU_df = pseU_df.rename(columns={'p_value_1': 'p_value_2'})\n",
    "    pseU2 = pd.merge(\n",
    "        pseU2,\n",
    "        pseU_df[['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'pseU_pos_1base', 'p_value_2']],\n",
    "        on=['chrom', 'snp_pos_1base', 'rsID', 'strand', 'EAF', 'p_value_2'],\n",
    "        how='left'\n",
    "    )\n",
    "    pseU2['m6A_pos_1base'] = ''\n",
    "\n",
    "    result_df = pd.concat([m6A1, m6A2, pseU1, pseU2, others], ignore_index=True)\n",
    "    result_df.to_csv(f'{outdir}/coloc_{chrom}_{strand}.csv', index=False)\n",
    "else:\n",
    "    print(f\"{chrom} {strand} has no significant colocalization\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 现在输入的是1个gene所有的share QTL，两两trait之间准备运行coloc\n",
    "def gene_2trait_combiner(geneID, gene_df, min_qtl_count, p_threshold, coloc_threshold):\n",
    "    trait_columns = ['tss', 'promoter', 'polyA', 'isoform', 'stability', 'm6A', 'pseU', 'ASE']\n",
    "    for col in trait_columns:\n",
    "        gene_df[col] = gene_df[f'p_value_{col}'].notna()\n",
    "    ## 下面用来生成所有两两trait组合的df\n",
    "    sub_dfs = {}\n",
    "    for trait1, trait2 in itertools.combinations(trait_columns, 2):\n",
    "        sub_df = gene_df[(gene_df[trait1] != False) & (gene_df[trait2] != False)]\n",
    "        if len(set(sub_df['snp_pos_1base'])) >= min_qtl_count :\n",
    "            sub_dfs[f\"{trait1}_{trait2}\"] = sub_df # 筛选2：两两trait组合的SNP数量大于阈值\n",
    "\n",
    "    if len(sub_dfs) == 0:\n",
    "        print(f'In {geneID},all trait combinations has less than {min_qtl_count} shared QTLs')\n",
    "        return None\n",
    "    else:\n",
    "        gene_coloc_res = pd.DataFrame(columns=[\n",
    "            \"chrom\",\"snp_pos_1base\",\"rsID\",\"strand\",\"EAF\",\n",
    "            \"p_value_1\",\"p_value_2\"\n",
    "            \"PPH4\",\"trait1\",\"trait2\"\n",
    "            ])\n",
    "        for trait1_trait2, sub_df in sub_dfs.items():\n",
    "            trait1, trait2 = trait1_trait2.split('_')\n",
    "            sub_df = sub_df[[\"chrom\" , \"snp_pos_1base\", \"rsID\", \"strand\", \"EAF\", f\"p_value_{trait1}\", f\"p_value_{trait2}\"]]\n",
    "            sub_df = sub_df.drop_duplicates()\n",
    "            if not {\"m6A\", \"pseU\"} & {trait1, trait2}:\n",
    "                coloc_res = R_coloc_prepare(sub_df, trait1, trait2, p_threshold, False, coloc_threshold)\n",
    "            else:\n",
    "                coloc_res = R_coloc_prepare(sub_df, trait1, trait2, p_threshold, True, coloc_threshold)\n",
    "            if coloc_res:\n",
    "                gene_coloc_res = pd.concat([gene_coloc_res, coloc_res], ignore_index=True)\n",
    "        if len(gene_coloc_res)==0:\n",
    "            return None\n",
    "        else:\n",
    "            gene_coloc_res['geneID'] = geneID\n",
    "            return gene_coloc_res # 返回chrom\tsnp_pos_1base\trsID\tstrand\tEAF p_value_1\tp_value_2 PPH4 trait1 trait2 geneID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 输入的是某两个trait之间的share QTL，处理一下重复的QTL，再运行colco\n",
    "def R_coloc_prepare(df, trait1, trait2, p_threshold, have_mod, coloc_threshold):\n",
    "    if (df[f\"p_value_{trait1}\"] > p_threshold).all() and (df[f\"p_value_{trait2}\"] > p_threshold).all():\n",
    "        print(f\"Both {trait1} and {trait2} are not significant in {geneID}\")\n",
    "        return None\n",
    "    else:\n",
    "        if have_mod:\n",
    "            sub_df['idx'] = list(range(1,len(sub_df)+1))\n",
    "            sub_df['snp_pos_1base'] = sub_df['snp_pos_1base'].astype(str)+\"_\"+sub_df['idx'].astype(str)\n",
    "            sub_df['rsID'] = sub_df['rsID']+\"_\"+sub_df['idx'].astype(str)\n",
    "            sub_df = sub_df.drop(columns=['idx'])\n",
    "            new_column_names = {\n",
    "                f\"p_value_{trait1}\": \"p_value_1\",\n",
    "                f\"p_value_{trait2}\": \"p_value_2\"\n",
    "            }\n",
    "            sub_df = sub_df.rename(columns=new_column_names)\n",
    "            coloc_res_df = R_coloc(sub_df, coloc_threshold)\n",
    "        else:\n",
    "            coloc_res_df = R_coloc(sub_df, coloc_threshold)\n",
    "        if coloc_res_df:\n",
    "            coloc_res_df['snp_pos_1base'] = coloc_res_df['snp_pos_1base'].apply(lambda x: x.split(\"_\")[0])\n",
    "            coloc_res_df['rsID'] = coloc_res_df['rsID'].apply(lambda x: x.split(\"_\")[0])\n",
    "            coloc_res_df['trait1'] = trait1\n",
    "            coloc_res_df['trait2'] = trait2\n",
    "            return coloc_res_df # 结果返回chrom\tsnp_pos_1base\trsID\tstrand\tEAF p_value_1\tp_value_2 PPH4 trait1 trait2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def R_coloc(df, coloc_threshold):\n",
    "    pandas2ri.activate()\n",
    "    robjects.r('library(coloc)')\n",
    "    r_df = pandas2ri.py2rpy(df) # 将 df 从 Pandas 数据框转换为 R 数据框，并赋值给 r_df\n",
    "    robjects.globalenv['df'] = r_df # 将 r_df 传递给 R 环境，并在 R 环境中将其命名为 df\n",
    "    robjects.globalenv['p1'] = p1\n",
    "    robjects.globalenv['p2'] = p2\n",
    "    robjects.globalenv['p12'] = p12\n",
    "    r_code = \"\"\"\n",
    "    result <- coloc.abf(\n",
    "        dataset1 = list(snp = df$rsID, pvalues = df$p_value_1, type = \"quant\", N = 104),\n",
    "        dataset2 = list(snp = df$rsID, pvalues = df$p_value_2, type = \"quant\", N = 104),\n",
    "        MAF = df$EAF,\n",
    "        p1 = {p1},\n",
    "        p2 = {p2},\n",
    "        p12 = {p12}\n",
    "    )\n",
    "    \"\"\"\n",
    "    robjects.r(r_code)\n",
    "    coloc_all_result = robjects.r['result']\n",
    "    # coloc_summary = coloc_all_result.rx2('summary')\n",
    "    coloc_results = coloc_all_result.rx2('results')\n",
    "    # priors = coloc_result.rx2('priors')\n",
    "    py_df = pd.DataFrame(robjects.conversion.rpy2py(coloc_results))\n",
    "    coloc_res_df = py_df[py_df['SNP.PP.H4'] > coloc_threshold]\n",
    "    if len(coloc_res_df) == 0:\n",
    "        return None\n",
    "    else:\n",
    "        coloc_res_df = coloc_res_df.rename(columns={'snp': 'rsID', 'SNP.PP.H4': 'PPH4'})\n",
    "        merged_df = pd.merge(df, coloc_res_df[['rsID', 'PPH4']], on='rsID', how='left')\n",
    "        return merged_df # 返回的结果就是所有PPH4大于阈值的行，并且最后一列是PPH4值"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "methy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
